### PR TITLE
Move logic after meta priming.

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/55620

Move menu item priming to after meta is primed. 

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by peterwilsoncc** [comment]
Am I correct in thinking the issue is the meta cache needs to be primed to reduce the queries from these two lines?

https://github.com/WordPress/wordpress-develop/blob/ec0f508a9ff1de37d9729785841a09250aa22462/src/wp-includes/nav-menu.php#L771-L772

A unit test or two would be dandy to make sure the code isn't later refactored to prime the menus prior to the meta data.

File: src/wp-includes/class-wp-query.php [hunk 1]
```diff
@@ -3465,6 +3461,10 @@ public function get_posts() {
 			$this->post_count = 0;
 			$this->posts      = array();
 		}
+
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-query.php]: Remove white space that resolve PHPCS

```suggestion

```

**Comment by spacedmonkey** [comment]
> A unit test or two would be dandy to make sure the code isn't later refactored to prime the menus prior to the meta data.

Unit tests added.

File: tests/phpunit/tests/post/nav-menu.php [hunk 2]
```diff
@@ -265,6 +265,77 @@ public function test_update_menu_item_cache_primes_terms() {
...
+					'menu-item-type'      => 'post_type',
+					'menu-item-object'    => 'post',
+					'menu-item-object-id' => $post_id,
+					'menu-item-status'    => 'publish',
+				)
+			);
+		}
+
+		wp_cache_delete_multiple( $menu_nav_ids, 'posts' );
+		wp_cache_delete_multiple( $post_ids, 'posts' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/nav-menu.php]:
```suggestion
		// Clear caches populated by `wp_update_nav_menu_item()` above.
		wp_cache_delete_multiple( $menu_nav_ids, 'posts' );
		wp_cache_delete_multiple( $post_ids, 'posts' );
```

File: tests/phpunit/tests/post/nav-menu.php [hunk 3]
```diff
@@ -265,6 +265,77 @@ public function test_update_menu_item_cache_primes_terms() {
...
+		$action = new MockAction();
+		add_filter( 'update_post_metadata_cache', array( $action, 'filter' ), 10, 2 );
+
+		wp_get_nav_menu_items( $this->menu_id );
+
+		$args = $action->get_args();
+		$this->assertSameSets( $menu_nav_ids, $args[0][1], '_prime_post_caches() was not executed.' );
+		$this->assertSameSets( $post_ids, $args[1][1], '_prime_post_caches() was not executed.' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/nav-menu.php:305]:
```suggestion
		$this->assertSameSets( $menu_nav_ids, $args[0][1], '_prime_post_caches() was not executed for navigation menu items.' );
		$this->assertSameSets( $post_ids, $args[1][1], '_prime_post_caches() was not executed for linked post objects.' );
```

I think this test is flakey, they only fail on trunk due to the order of operations:

* db queries: 6 on both trunk and this branch
* posts cached: yes. runs first on trunk, second on this branch
* nav items cached: yes. Runs second on trunk, first on this branch

File: tests/phpunit/tests/post/nav-menu.php [hunk 4]
```diff
@@ -265,6 +265,77 @@ public function test_update_menu_item_cache_primes_terms() {
...
+		}
+		wp_cache_delete_multiple( $menu_nav_ids, 'posts' );
+		wp_cache_delete_multiple( $term_ids, 'terms' );
+		$action_terms = new MockAction();
+		add_filter( 'update_term_metadata_cache', array( $action_terms, 'filter' ), 10, 2 );
+
+		$action_posts = new MockAction();
+		add_filter( 'update_post_metadata_cache', array( $action_posts, 'filter' ), 10, 2 );
+
+		wp_get_nav_menu_items( $this->menu_id );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/nav-menu.php:340]: Same in this test, both trunk and this branch show six database queries.

**Review by peterwilsoncc — COMMENTED** [review]
The new tests don't appear to test the change (or I am doing something silly) -- the number of database queries for `wp_get_nav_menu_items()` is the same on both trunk and this branch.

Testing the number of queries are as expected is a good thing but I am not clear this change affects them.

**Comment by spacedmonkey** [comment]
@peterwilsoncc Good call out. I have improves the tests.

File: tests/phpunit/tests/post/nav-menu.php [hunk 5]
```diff
@@ -265,6 +265,90 @@ public function test_update_menu_item_cache_primes_terms() {
...
+		wp_cache_delete_multiple( $post_ids, 'posts' );
+		wp_cache_delete_multiple( $post_ids, 'post_meta' );
+
+		$action = new MockAction();
+		add_filter( 'update_post_metadata_cache', array( $action, 'filter' ), 10, 2 );
+
+		$num_queries = get_num_queries();
+		wp_get_nav_menu_items( $this->menu_id );
+
+		$this->assertSame( $num_queries + 7, get_num_queries(), 'Only does 7 database queries when running wp_get_nav_menu_items.' );
```
Comment (peterwilsoncc) [inline tests/phpunit/tests/post/nav-menu.php]:
```suggestion
		$start_num_queries = get_num_queries();
		wp_get_nav_menu_items( $this->menu_id );
		$queries_made = get_num_queries() - $start_num_queries;

		$this->assertSame( 7, $queries_made, 'Only does 7 database queries when running wp_get_nav_menu_items.' );
```

When running on trunk I get `Failed asserting that 532 is identical to 530.` which is much less helpful than `Failed asserting that 9 is identical to 7.`

:1234: This same change applies below.

**Review by peterwilsoncc — COMMENTED** [review]
Changes look good and are now failing on trunk. I can see what is happeneing now, thank you.

I've made a suggestion for clarity when the tests fail but feel free to push back.

**Comment by spacedmonkey** [comment]
Committed https://github.com/WordPress/wordpress-develop/commit/fd60449f92180fd73f4d805d9712dba398f70a9d

**Pull request closed without merging**